File: DSA/src/Questions/Graph/PathDecisions/FloydWarshall.java
package Questions.Graph.PathDecisions;

public class FloydWarshall {
    public void shortestDistance(int[][] matrix) {
        int n = matrix.length;

        // Initialize matrix: replace -1 with a large value and set diagonal to 0
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == -1) {
                    matrix[i][j] = Integer.MAX_VALUE;
                }
                if (i == j) {
                    matrix[i][j] = 0;
                }
            }
        }

        // Floyd-Warshall algorithm for finding shortest distances
        for (int k = 0; k < n; k++) {
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    // Update shortest distance if going through vertex k is shorter
                    if (matrix[i][k] != Integer.MAX_VALUE && matrix[k][j] != Integer.MAX_VALUE) {
                        matrix[i][j] = Math.min(matrix[i][j], matrix[i][k] + matrix[k][j]);
                    }
                }
            }
        }

        // Restore original representation: replace large values with -1
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == Integer.MAX_VALUE) {
                    matrix[i][j] = -1;
                }
            }
        }
    }

    public static void main(String[] args) {
        // Example usage
        int[][] matrix = {
                { 0, 5, -1, 10 },
                { -1, 0, 3, -1 },
                { -1, -1, 0, 1 },
                { -1, -1, -1, 0 }
        };

        FloydWarshall fw = new FloydWarshall();
        fw.shortestDistance(matrix);

        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                System.out.print(matrix[i][j] + " ");
            }
            System.out.println();
        }
    }
}

File: DSA/Python/Questions/Graph/PathDecisions/FloydWarshall.py
"""
# Question: Floyd Warshall Algorithm Implementation
# Finds all pairs shortest paths in a weighted graph

# Time Complexity: O(V^3)
# Space Complexity: O(V^2)

# Algorithm:
# 1. Initialize distance matrix
# 2. Consider all vertices as intermediate
# 3. Update shortest paths through intermediates
"""


class FloydWarshall:
    def all_pairs_shortest_paths(
        self, n: int, edges: list[list[int]]
    ) -> list[list[int]]:
        # Initialize distances
        dist = [[float("inf")] * n for _ in range(n)]
        for i in range(n):
            dist[i][i] = 0

        # Set direct edges
        for u, v, w in edges:
            dist[u][v] = w

        # Floyd-Warshall algorithm
        for k in range(n):
            for i in range(n):
                for j in range(n):
                    if dist[i][k] != float("inf") and dist[k][j] != float("inf"):
                        dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])

        return dist


def main():
    solution = FloydWarshall()
    edges = [[0, 1, 4], [0, 2, 3], [1, 2, 1], [2, 3, 2]]
    print(solution.all_pairs_shortest_paths(4, edges))


if __name__ == "__main__":
    main()
